project-koku/koku
# COST-239: Add provider_uuid to the masu endpoint crawl_account_hierachy

This pull request is to address some of @lcouzens comments regarding the initial pull request: https://github.com/project-koku/koku/pull/2218

Jira Issue: https://issues.redhat.com/browse/COST-239

**What this pr is trying to fix:**
In general, Masu endpoints wont be accessible to anyone since its part of the masu spec and you can only get to that locally. However,  IQE can do it since they do port forwarding to the specific pods. The way the `crawl_account_hierarchy` endpoint currently works when Luke proxies to the pod hit the endpoint directly and that actually runs for all schema's. We need to lock it down to the schema or provider uuid so that it won't run the crawler for everyone when IQE hits the endpoint directly on ci. 

Additionally I updated the openapi.spec.

--------

#### How to Test

The first step to testing is to reinit the db with sources.

  Then add an aws source that can be crawled with:

  ``` make aws-source aws_name=AWS-SOURCE-001 bucket=cost-usage-bucket ```  

Then add some nise sources: 
``` make load-test-customer-data ``` 

<wait for all the data to be loaded>  

Now we need to look up the provider_uuid or source_uuid for the 1st aws source we added. We can do that by looking at this endpoint:
     http://127.0.0.1:8000/api/cost-management/v1/sources/  

We now need to pass that uuid to the crawl account heirarchy endpoint:
       http://127.0.0.1:5000/api/cost-management/v1/crawl_account_hierarchy/?provider_uuid=<provider_uuid>  

Now check the logs:
```
 koku_worker       | [2020-07-06 15:15:18,124: INFO/ForkPoolWorker-1] [masu.celery.tasks.crawl_account_hierarchy(668f14cb-fed7-47db-be5d-290b1eb88b1a via 668f14cb-fed7-47db-be5d-290b1eb88b1a)] Account hierarchy crawler found 1 accounts to scan
koku_worker       | [2020-07-06 15:15:18,124: INFO/ForkPoolWorker-1] [masu.celery.tasks.crawl_account_hierarchy(668f14cb-fed7-47db-be5d-290b1eb88b1a via 668f14cb-fed7-47db-be5d-290b1eb88b1a)] Starting account hierarchy crawler for type AWS with provider_uuid (e624b16d-78b6-4a64-8a44-5c7bc9877ec1)
masu_server       | [2020-07-06 15:15:18,218] INFO None {'method': 'GET', 'path': '/api/cost-management/v1/crawl_account_hierarchy/?provider_uuid=e624b16d-78b6-4a64-8a44-5c7bc9877ec1', 'status': 200, 'request_id': None, 'account': None, 'username': None, 'is_admin': False}
masu_server       | [2020-07-06 15:15:18,222] INFO None "GET /api/cost-management/v1/crawl_account_hierarchy/?provider_uuid=e624b16d-78b6-4a64-8a44-5c7bc9877ec1 HTTP/1.1" 200 6639
masu_server       | [2020-07-06 15:15:18,337] INFO None "GET /api/cost-management/static/rest_framework/css/bootstrap-tweaks.css HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,345] INFO None "GET /api/cost-management/static/rest_framework/css/default.css HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,348] INFO None "GET /api/cost-management/static/rest_framework/css/prettify.css HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,353] INFO None "GET /api/cost-management/static/rest_framework/css/bootstrap.min.css HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,370] INFO None "GET /api/cost-management/static/rest_framework/js/csrf.js HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,371] INFO None "GET /api/cost-management/static/rest_framework/js/jquery-3.4.1.min.js HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,383] INFO None "GET /api/cost-management/static/rest_framework/js/bootstrap.min.js HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,395] INFO None "GET /api/cost-management/static/rest_framework/js/ajax-form.js HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,400] INFO None "GET /api/cost-management/static/rest_framework/js/default.js HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,403] INFO None "GET /api/cost-management/static/rest_framework/js/prettify-min.js HTTP/1.1" 304 0
masu_server       | [2020-07-06 15:15:18,430] INFO None "GET /api/cost-management/static/rest_framework/img/grid.png HTTP/1.1" 304 0
koku_worker       | [2020-07-06 15:15:18,606] INFO 668f14cb-fed7-47db-be5d-290b1eb88b1a Starting aws organizations session for crawler.
koku_worker       | [2020-07-06 15:15:18,616] INFO 668f14cb-fed7-47db-be5d-290b1eb88b1a Tree for 2020-07-05 to 2020-07-05
koku_worker       | [2020-07-06 15:15:19,315] INFO 668f14cb-fed7-47db-be5d-290b1eb88b1a Obtained the root identifier: r-f22a
koku_worker       | [2020-07-06 15:15:19,324] INFO 668f14cb-fed7-47db-be5d-290b1eb88b1a Saving account or org unit: unit_name=Root, unit_id=r-f22a, unit_path=r-f22a, account_alias=None, level=0
koku_worker       | [2020-07-06 15:15:19,324] INFO 668f14cb-fed7-47db-be5d-290b1eb88b1a Obtaining accounts for organizational unit: r-f22a
koku_worker       | [2020-07-06 15:15:19,691] INFO 668f14cb-fed7-47db-be5d-290b1eb88b1a Saving account or org unit: unit_name=Root, unit_id=r-f22a, unit_path=r-f22a, account_alias={ id : 1, account_id : 589173575009, account_alias : CH Red Hat }, level=0
koku_worker       | [2020-07-06 15:15:19,904: INFO/ForkPoolWorker-1] [masu.celery.tasks.crawl_account_hierarchy(668f14cb-fed7-47db-be5d-290b1eb88b1a via 668f14cb-fed7-47db-be5d-290b1eb88b1a)] Account hierarchy crawler finished. 1 processed and 0 skipped
```

 You will notice that the crawler only found one account to scan.   

Now lets hit the same endpoint again with a fake provider uuid:  
     http://127.0.0.1:5000/api/cost-management/v1/crawl_account_hierarchy/?provider_uuid=abc123  

You should see the following error:
 ```
 "Error": "The provider_uuid abc123 does not exist."
 ```  

And finally make sure that the provider_uuid is now required:
     http://127.0.0.1:5000/api/cost-management/v1/crawl_account_hierarchy/
